libs/@blockprotocol/graph/src/internal/mutate-subgraph.ts

```diff
@@ -50,103 +54,128 @@ export const addKnowledgeGraphEdgeToSubgraphByMutation = (
 };
 
 /**
- * Looking to build a subgraph? You probably want `import { buildSubgraph } from @blockprotocol/graph/stdlib`
+ * Looking to build a subgraph? You probably want {@link buildSubgraph} from `@blockprotocol/graph/stdlib`
  *
- * This MUTATES the given `Subgraph` by adding the given outwardEdge to the entity at the specified atTime.
+ * This MUTATES the given {@link Subgraph} by adding a given list of entities to the vertices, creating any link edges
+ * that are **directly implied** by any link entities in the list (see note below).
  * Mutating a Subgraph is unsafe in most situations – you should know why you need to do it.
  *
- * *Note*: This only adds edges as implied by the given entities, if the `Subgraph` is invalid at the time of method
- * call (e.g. by missing link endpoints), this will not loop through the vertex set to finish incomplete edges.
+ * *Note*: This only adds edges as implied by the given entities, if the {@link Subgraph} is invalid at the time of
+ * method call (e.g. by missing link endpoints), this will not loop through the vertex set to finish incomplete edges.
  *
- * @param subgraph – the subgraph to mutate by adding the provided entities
- * @param entities – the entity to add to the provided subgraph
+ * @param {Subgraph} subgraph – the subgraph to mutate by adding the provided entities
+ * @param {Entity[]} entities – the entity to add to the provided subgraph
  */
-export const addEntitiesToSubgraphByMutation = (
-  subgraph: Subgraph,
-  entities: Entity[],
+export const addEntitiesToSubgraphByMutation = <Temporal extends boolean>(
+  subgraph: Subgraph<Temporal>,
+  entities: Entity<Temporal>[],
 ) => {
+  /*
+   * @todo This assumes that the left and right entity ID of a link entity is static for its entire lifetime, that is
+   *   not necessarily going to continue being the case
```

Comment:
I was going to raise this, it feels pretty consequential for HASH as well

Reply:
I'm not sure, that's part of the issue. 

I think it's presently somewhat undefined/unexplored behaviour that's already present within the constraints of our system as we have specified them. We have encoded within our own implementations (but not the definitions of the primitives and their operations) that a link's endpoints cannot change. But the primitives do not encode that.

I think we maybe just need a task to look into this?

Reply:
The primitives don't, but if we take the message signatures in `@blockprotocol/graph` as representing what will be defined the spec, they do for now encode that the left and right entity can only be set on creation and never again.

We can create a task if we have some way of capturing why we would want to revisit that.

Reply:
For completeness, I was actually considering the message signatures and such as part of the "primitives" (I appreciate that's perhaps using the term too loosely).

I think we still have a lack of clarity within the current system because of these scenarios
1. You call `createEntity` for a link entity, set its leftEntity and rightEntity
   1. You call `updateEntity` and change its entity type to something that _is not_ a link entity type
   2. What happens? 
      1. Do we continue to return left and right entity IDs in the metadata?
      2. Do we delete the left and right entity information?   
      3. What should backends do?
    3. You now call `updateEntity` again and change its entity type to something that _is_ a link entity type (perhaps a completely different type)
        1. What do you do about the various options above?
 2.  You call `createEntity` for a _non_-link entity, you do **not** set its leftEntity and rightEntity
     1. You now call `updateEntity` and change its entity type to something that _is_ a link entity type
     2. What happens?  
          1. Do you continue to have an entity that breaks the invariants where we say all link entities must have left and right endpoints?
          2. Do you change `updateEntity` to allow setting of left and right endpoints?
          3. Do you somehow block the `updateEntity` operation by not allowing its entity type to be changed to something that's a link? There's nothing in the messages, documentation, or primitives that suggest anything EAs should do in this situation.

I think there are some viable approaches (including starting to more concretely specify validation of mutation methods, which we don't currently do to my knowledge), but these are things we need to visit rather than revisit, in my eyes.

Reply:
Hmm, fair enough. I was thinking that it is impossible to set the left/right entity after creation, but you're right that this creates issues with type switching that we need to figure out what to do about.